packages/sdk/client/src/services/socket.ts

```diff
@@ -31,13 +36,25 @@ export const fromSocket = async (url: string): Promise<ClientServicesProvider> =
       const { WebsocketRpcClient } = await import('@dxos/websocket-rpc');
       dxRpcClient = new WebsocketRpcClient({
         url,
+        authenticationToken,
         requested: clientServiceBundle,
         exposed: {},
         handlers: {},
       });
+
+      dxRpcClient.error.on(async (error) => {
+        log.warn('websocket rpc client error', { error });
+        // Browsers do not include the error message in the event object, so we cannot discern 401 errors from other errors.
+        if (error.message.includes('401')) {
+          log.warn('websocket authentication failed');
+        }
+        fatal.emit(error);
```

Comment:
I think that this `fatal` event might be leaking internal implementation details of the socket services provider into client. The way I would expect this to work is that if there's an error initially that opening would throw [this trigger](https://github.com/dxos/dxos/pull/6345/files#diff-785917a9d319e96d99ca9d7921879288a3b0921055c1cb6e0db85a7cc63d2d1bR312) wouldn't be needed.

If this is more for if the websocket fails after initialization, then I think we can just extend the `closed` event with a `fatal` flag that tells client not to try to restart

Reply:
yeah, I suspected that might be the case, and your solution is probably better than making initializing the ClientServicesProvider synchronous.